### PR TITLE
rt_sigprocmask syscall implementation

### PR DESCRIPTION
**Pull request by Lurk**
This pull request adds rt_sigprocmask syscall handler (https://man7.org/linux/man-pages/man2/sigprocmask.2.html). 

Before adding signal filtering, I want to discuss the general approach. 

Things that bother me: 

I feel like the `how` argument should be an enum, but how to handle `EINVAL` in that case?

And the `length` argument doesn't make any sense to me. According to the manual, it is `sizeof(kernel_sigset_t)`, and according to kernel sources, `kernel_sigset_t` is 1024 bits. So it always, according to POSIX, should be 128 bytes, unless we will run on something exotic like PDP-6/10 or DSP. If you can point me to some document or discussion to make sense of it, I would be grateful. 

Also `set` and `oldset` are 1024 bits in length. I am using a slightly modified BitMap from kerla_utils, but it is suboptimal. It is [u8;128], which means we have 128 iterations on `SIG_BLOCK` and `SIG_UNBLOCK`. The solution can be generic over the type BitMap. Or create a new LongBitMap, which will use u128 under the hood. Or, maybe we can use or take inspiration from something like https://docs.rs/bitmaps/latest/bitmaps/. But I have a feeling that this is a task by itself, and should be discussed in a separate issue/PR.

### CODE REVIEW

File: libs/kerla_utils/bitmap.rs [hunk 1]
```diff
@@ -51,6 +50,30 @@ impl<const SZ: usize> BitMap<SZ> {
...
+    }
+
+    pub fn assign_or(&mut self, rhs: [u8; SZ]) {
+        let mut iter = rhs.into_iter();
+        for byte in &mut self.0 {
+            *byte |= iter.next().unwrap_or(0);
+        }
+    }
+
+    pub fn assign_material_nonimplication(&mut self, rhs: [u8; SZ]) {
```
Comment (nuta) [inline libs/kerla_utils/bitmap.rs:65]: How about `and_not`?

```suggestion
    pub fn and_not(&mut self, rhs: [u8; SZ]) {
```

File: libs/kerla_utils/bitmap.rs [hunk 2]
```diff
@@ -51,6 +50,30 @@ impl<const SZ: usize> BitMap<SZ> {
...
+        let mut iter = rhs.into_iter();
+        for byte in &mut self.0 {
+            *byte &= !iter.next().unwrap_or(0);
+        }
```
Comment (nuta) [inline libs/kerla_utils/bitmap.rs:69]: Let's access `rhs` without using iterators and add a doc comment to note that this method would panic if `SZ != rhs.len()`.

```suggestion
        for (i, byte) in self.0.iter_mut().enumerate() {
            *byte &= !rhs[i];
        }
```

File: libs/kerla_utils/bitmap.rs [hunk 3]
```diff
@@ -51,6 +50,30 @@ impl<const SZ: usize> BitMap<SZ> {
...
+
+    pub fn assign(&mut self, rhs: [u8; SZ]) {
+        self.0 = rhs;
+    }
+
+    pub fn assign_or(&mut self, rhs: [u8; SZ]) {
+        let mut iter = rhs.into_iter();
+        for byte in &mut self.0 {
+            *byte |= iter.next().unwrap_or(0);
+        }
```
Comment (nuta) [inline libs/kerla_utils/bitmap.rs:62]: Ditto: Let's access `rhs` without using iterators.

File: kernel/process/process.rs [hunk 4]
```diff
@@ -351,6 +356,38 @@ impl Process {
         self.signals.lock().is_pending()
     }
 
+    /// Sets signal mask
+    pub fn rt_sigprocmask(
```
Comment (nuta) [inline kernel/process/process.rs:360]: How about using a more intuitive name like following?

```suggestion
    pub fn set_signal_mask(
```

File: kernel/process/process.rs [hunk 5]
```diff
@@ -351,6 +356,38 @@ impl Process {
...
+        let mut sigset = self.sigset.lock();
+
+        if let Some(old) = oldset {
+            if let Err(_) = old.write_bytes(sigset.as_slice()) {
+                return Err(Error::new(Errno::EFAULT));
+            }
+        }
+
+        if let Some(new) = set {
+            if let Ok(new_set) = new.read::<[u8; 128]>() {
```
Comment (nuta) [inline kernel/process/process.rs:376]: Nit: I guess you can use `?` here.

```suggestion
            let new_set = new.read::<[u8; 128]>()?;
```

**Comment by nuta** [comment]
Excellent! This is one of major missing features in Kerla 🎉 

> I feel like the how argument should be an enum, but how to handle EINVAL in that case?

How about using `match` in `sys_rt_sigprocmask` to do that, like you did in `Process::rt_sigprocmask`? The number of possible `how` looks small enough to me.

```rust
let how = match how {
    0 => SignalMask::Block,
     _ => return Err(Errno::EINVAL.into());
};

current_process().rt_sigprocmask(how, set, oldset, length)
```

> And the length argument doesn't make any sense to me. According to the manual, it is sizeof(kernel_sigset_t), and according to kernel sources, kernel_sigset_t is 1024 bits. So it always, according to POSIX, should be 128 bytes, unless we will run on something exotic like PDP-6/10 or DSP. If you can point me to some document or discussion to make sense of it, I would be grateful.

I agree we can go with 1024 bits. Let's add `debug_warn!` if the length is not 1024 just in case.

Regarding the bitmap stuff, I think there're some possible optimizations too. The crate you mentioned looks neat and perhaps we replace our bitmap implementation with it. Let's discuss this topic in a separate issue later 😃

**Review by nuta — CHANGES_REQUESTED** [review]
Could you have a look at my review?

File: libs/kerla_utils/bitmap.rs (hunk 1 above)
Comment (Lurk) [reply libs/kerla_utils/bitmap.rs]: Done

File: libs/kerla_utils/bitmap.rs (hunk 2 above)
Comment (Lurk) [reply libs/kerla_utils/bitmap.rs]: Done, but I am not sure if I did comment part correctly :)

File: libs/kerla_utils/bitmap.rs (hunk 3 above)
Comment (Lurk) [reply libs/kerla_utils/bitmap.rs]: Done

File: libs/kerla_utils/bitmap.rs [hunk 6]
```diff
@@ -51,6 +50,30 @@ impl<const SZ: usize> BitMap<SZ> {
...
+        let mut iter = rhs.into_iter();
+        for byte in &mut self.0 {
+            *byte &= !iter.next().unwrap_or(0);
+        }
+    }
+}
+
+impl<const SZ: usize> Clone for BitMap<SZ> {
+    fn clone(&self) -> Self {
+        BitMap(self.0)
```
[the remark this answers is not in the document]
Comment (Lurk) [reply libs/kerla_utils/bitmap.rs]: No, I don't know why I did not try it in the first place.

File: kernel/process/process.rs (hunk 4 above)
Comment (Lurk) [reply kernel/process/process.rs]: Done

File: kernel/process/process.rs (hunk 5 above)
Comment (Lurk) [reply kernel/process/process.rs]: Done

File: kernel/process/process.rs [hunk 7]
```diff
@@ -351,6 +356,38 @@ impl Process {
...
+                    1 => sigset.assign_material_nonimplication(new_set),
+                    2 => sigset.assign(new_set),
+                    _ => return Err(Error::new(Errno::EINVAL)),
+                }
+            } else {
+                return Err(Error::new(Errno::EFAULT));
+            }
+        }
+
+        Ok(0)
```
[the remark this answers is not in the document]
Comment (Lurk) [reply kernel/process/process.rs]: Done

File: kernel/process/process.rs [hunk 8]
```diff
@@ -351,6 +356,38 @@ impl Process {
...
     }
 
+    /// Sets signal mask
+    pub fn rt_sigprocmask(
+        &self,
+        how: usize,
+        set: Option<UserVAddr>,
+        oldset: Option<UserVAddr>,
+        _length: usize,
+    ) -> Result<isize> {
```
[the remark this answers is not in the document]
Comment (Lurk) [reply kernel/process/process.rs]: Done

File: libs/kerla_utils/bitmap.rs [hunk 9]
```diff
@@ -71,4 +89,18 @@ mod tests {
...
+
+    #[test]
+    fn or() {
+        let mut bitmap = BitMap::from_array([0b0100_0010, 0b1000_0001]);
+        bitmap.assign_or([0b0010_0100, 0b1010_0110]);
+        assert_eq!(bitmap.as_slice(), &[0b0110_0110, 0b1010_0111]);
+    }
+
+    #[test]
+    fn material_nonimplication() {
```
Comment (nuta) [inline libs/kerla_utils/bitmap.rs:101]:
```suggestion
    fn assign_and_not() {
```

File: kernel/process/process.rs [hunk 10]
```diff
@@ -359,17 +390,20 @@ impl Process {
         // TODO: sigmask
```
Comment (nuta) [inline kernel/process/process.rs:390]: Nit: Remove this TODO comment as you've implemented it ;)

```suggestion
```

**Review by nuta — COMMENTED** [review]
Generally looks good to me! Just leave some nitpicks.

**Comment by nuta** [comment]
> I just checked, every time rt_sigprocmask is called during our integration tests, the length argument equals 8. I am lost :)
> It is like long long int in C is 128 bit, but google says it is 64 bits. Anyways, should I debug_warn if length is not 8?

It seems I was confused with the difference between `sigset_t` and `kernel_sigset_t`. `sigset_t` is 1024 bits-sized type but apparently `kernel_sigset_t` is 8 bytes at lease in x64.

Please check if the length is `8` instead of 1024 bits in the system call handler.

**Comment by Lurk** [comment]
add debug_warn! if length argument is not equal 8

**Comment by Lurk** [comment]
> Please check if the length is 8 instead of 1024 bits in the system call handler.
Done

I will create a ticket for bitmap stuff later.

**Review by nuta — APPROVED** [review]

**Comment by nuta** [comment]
I'll update `Documentation/compatibility.md` by myself. Let me merge this PR as it is.

**Comment by nuta** [comment]
Thanks a lot, @Lurk!

**Comment by Lurk** [comment]
@nuta somehow I missed Clippy warning :(

```sh
error: redundant pattern matching, consider using `is_err()`
  --> kernel/syscalls/rt_sigprocmask.rs:27:16
   |
27 |         if let Err(_) = current_process().set_signal_mask(how, set, oldset, length) {
   |         -------^^^^^^-------------------------------------------------------------- help: try this: `if current_process().set_signal_mask(how, set, oldset, length).is_err()`
   |
   = note: `-D clippy::redundant-pattern-matching` implied by `-D warnings`
```
Should I open a new pull request?

**Pull request merged**